### PR TITLE
1.20.4 update and ping handler in bot/basic

### PR DESCRIPTION
**Pull request by qwqtoday**

### CODE REVIEW

**Comment by Tnze** [comment]
Thanks for PR!
Did you test it on the 1.20.4 vanilla server?

**Comment by qwqtoday** [comment]
There seem to be issues with bot/chat after this update.

**Comment by qwqtoday** [comment]
> Thanks for PR! Did you test it on the 1.20.4 vanilla server?

no not yet, I will try it now

**Comment by qwqtoday** [comment]
> Thanks for PR! Did you test it on the 1.20.4 vanilla server?

I ran daze and only seeing this error
```
handle packet ClientboundPlayerChat error: scanning packet field[6] error: invalid character '\x00' looking for beginning of value
```

**Comment by Tnze** [comment]
At least the bot can join the game. The login processing is good.

The error message telling us the ClientboundPlayerChat packet format is updated. This can be fixed in other commit later.

**Review by Tnze — APPROVED** [review]
Looks good to me

**Comment by qwqtoday** [comment]
Apparently text can be just a string or a nbt tag after protocol version 492, so if we want to fix it we can check the type id first and read the string

**Comment by qwqtoday** [comment]
The player chat packet had changed a bit after protocol version 492
We just have to fix the handler in bot/chat now

File: chat/message.go
Comment (Tnze) [inline chat/message.go:1]: This change is wrong.

According to <https://wiki.vg/Protocol#Type:Text_Component>, a `Text Component` is encode in a single NBT. So it's corresponding to a single `nbt.Unmarshal()` call. The distinguish logic should be placed in independent `UnmarshalNBT` method just like the `UnmarshalJSON()` above.

File: chat/message.go [hunk 1]
```diff
@@ -133,13 +134,23 @@ func (m *Message) UnmarshalJSON(raw []byte) (err error) {
...
-	n, err := code.ReadFrom(r)
+	n, err = code.ReadFrom(r)
 	if err != nil {
 		return n, err
 	}
-	err = json.Unmarshal([]byte(code), m)
+	if byte(msgType) == nbt.TagString {
+		m.Text = string(code)
+	} else if byte(msgType) == nbt.TagCompound {
+		err = json.Unmarshal([]byte(code), m)
```
Comment (Tnze) [inline chat/message.go:152]: And this shouldn't be `json.Unmarshal()`

**Review by Tnze — CHANGES_REQUESTED** [review]

File: chat/message.go
Comment (qwqtoday) [reply chat/message.go:1]: It was using json.Unmarshal() originally, and the fields in the Message struct uses the json tag

**Comment by Tnze** [comment]
I will revert two commit and merge this PR. We could solve the chat.Message problem later.

**Comment by Tnze** [comment]
The TextComponent things is fixed

**Pull request merged**